File: bindings/android/src/main/java/com/nativeview/NativeViewWindow.java
package com.nativeview;

import com.nativeview.jni.NativeViewJNI;

/**
 * A native-backed window handle; mirrors the desktop Java binding surface while WebView hosting
 * remains in the Android {@link com.nativeview.jni.NativeViewHost} implementation.
 */
public class NativeViewWindow {

    private final long nativeHandle;

    NativeViewWindow(long nativeHandle) {
        this.nativeHandle = nativeHandle;
    }

    /** Opaque native address; zero when invalid. */
    public long getNativeHandle() {
        return nativeHandle;
    }

    public void show() {
        if (nativeHandle == 0L) {
            return;
        }
        NativeViewJNI.nativeShow(nativeHandle);
    }

    public void hide() {
        if (nativeHandle == 0L) {
            return;
        }
        NativeViewJNI.nativeHide(nativeHandle);
    }

    public void load_url(String url) {
        if (nativeHandle == 0L || url == null) {
            return;
        }
        NativeViewJNI.nativeLoadUrl(nativeHandle, url);
    }

    public void load_html(String html, String baseUrl) {
        if (nativeHandle == 0L) {
            return;
        }
        NativeViewJNI.nativeLoadHtml(nativeHandle, html != null ? html : "", baseUrl);
    }

    public void eval_js(String js) {
        if (nativeHandle == 0L) {
            return;
        }
        NativeViewJNI.nativeEvalJs(nativeHandle, js != null ? js : "");
    }

    public void set_title(String title) {
        if (nativeHandle == 0L || title == null) {
            return;
        }
        NativeViewJNI.nativeSetTitle(nativeHandle, title);
    }

    public void set_size(int width, int height) {
        if (nativeHandle == 0L) {
            return;
        }
        NativeViewJNI.nativeSetSize(nativeHandle, width, height);
    }

    public void on_message(MessageCallback callback) {
        if (nativeHandle == 0L) {
            return;
        }
        NativeViewJNI.nativeOnMessage(nativeHandle, callback);
    }

    public void on_ready(ReadyCallback callback) {
        if (nativeHandle == 0L) {
            return;
        }
        NativeViewJNI.nativeOnReady(nativeHandle, callback);
    }

    public void on_close(CloseCallback callback) {
        if (nativeHandle == 0L) {
            return;
        }
        NativeViewJNI.nativeOnClose(nativeHandle, callback);
    }

    /**
     * Sends an event to the page via {@code nv_send}. {@code event} must be non-null; use {@code
     * ""} for IPC-style payloads consumed by {@code window.__nv._emit}.
     */
    public void emit(String event, String json) {
        if (nativeHandle == 0L || event == null) {
            return;
        }
        NativeViewJNI.nativeEmit(nativeHandle, event, json != null ? json : "null");
    }
}

File: bindings/ios/Sources/NativeViewIOS/Window/NativeViewWindow.swift
// SPDX-License-Identifier: Apache-2.0

import UIKit

/// Native-backed window host for a nativeview page (parity with desktop bindings); embed via ``viewController()``.
public final class NativeViewWindow: @unchecked Sendable {
    weak var app: NativeViewApp?

    private let rootViewController: NativeViewViewController
    private let lock = NSLock()
    private var messageCallback: MessageCallback?
    private var readyCallback: ReadyCallback?
    private var closeCallback: CloseCallback?
    private var config: NativeViewConfig

    init(app: NativeViewApp?, config: NativeViewConfig) {
        self.app = app
        self.config = config
        let bridge = app?.bridge ?? NVBridge()
        let vc = NativeViewViewController(bridge: bridge, config: config)
        self.rootViewController = vc
        vc.hostWindow = self
        vc.webView.hostWindow = self
    }

    /// Shows the hosted view (sets `isHidden` on the root view once loaded).
    public func show() {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.rootViewController.loadViewIfNeeded()
            self.rootViewController.hostViewIsHidden = false
        }
    }

    /// Hides the hosted view.
    public func hide() {
        DispatchQueue.main.async { [weak self] in
            self?.rootViewController.hostViewIsHidden = true
        }
    }

    public func loadUrl(_ url: String) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.rootViewController.loadViewIfNeeded()
            let wv = self.rootViewController.webView
            guard let u = URL(string: url) else { return }
            // `load(_:)` does not grant the WebContent process read access to `file:` URLs; use `loadFileURL` instead.
            if u.isFileURL {
                let readAccess = u.deletingLastPathComponent()
                wv.loadFileURL(u, allowingReadAccessTo: readAccess)
            } else {
                wv.load(URLRequest(url: u))
            }
        }
    }

    public func loadHtml(_ html: String, baseUrl: String = "about:blank") {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.rootViewController.loadViewIfNeeded()
            let wv = self.rootViewController.webView
            wv.loadHTMLString(html, baseURL: URL(string: baseUrl))
        }
    }

    /// When set (e.g. by `@testable` unit tests), ``evalJs(_:)`` forwards here instead of evaluating in WKWebView.
    internal var evalJsCapture: ((String) -> Void)?

    public func evalJs(_ js: String) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            if let cap = self.evalJsCapture {
                cap(js)
                return
            }
            self.rootViewController.loadViewIfNeeded()
            let wv = self.rootViewController.webView
            wv.evaluateJavaScript(js, completionHandler: nil)
        }
    }

    public func setTitle(_ title: String) {
        config.title = title
        DispatchQueue.main.async { [weak self] in
            self?.rootViewController.title = title
        }
    }

    public func setSize(width: Int, height: Int) {
        config.width = width
        config.height = height
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.rootViewController.preferredContentSize = CGSize(width: width, height: height)
        }
    }

    public func onMessage(_ callback: @escaping MessageCallback) {
        lock.lock()
        defer { lock.unlock() }
        messageCallback = callback
    }

    public func onReady(_ callback: @escaping ReadyCallback) {
        lock.lock()
        defer { lock.unlock() }
        readyCallback = callback
    }

    public func onClose(_ callback: @escaping CloseCallback) {
        lock.lock()
        defer { lock.unlock() }
        closeCallback = callback
    }

    /// Sends a push event into the page (`window.__nv._emit` when nativeview.js is loaded).
    public func emit(event: String, json: String) {
        let js =
            "(function(){try{if(window.__nv&&typeof window.__nv._emit==='function'){window.__nv._emit(\(NVJavaScriptStringEncoding.jsStringLiteral(event)), \(NVJavaScriptStringEncoding.jsStringLiteral(json)));}}catch(e){}})()"
        evalJs(js)
    }

    /// Returns the root view controller to embed in your hierarchy (UINavigationController, UITabBarController, etc.).
    public func viewController() -> UIViewController {
        rootViewController
    }

    func notifyMessage(event: String, json: String) {
        lock.lock()
        let cb = messageCallback
        lock.unlock()
        cb?(event, json)
    }

    func notifyReady() {
        lock.lock()
        let cb = readyCallback
        lock.unlock()
        cb?()
    }

    func notifyClose() {
        lock.lock()
        let cb = closeCallback
        lock.unlock()
        cb?()
    }

    func tearDown() {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.rootViewController.detachWebView()
        }
    }
}
